cpp/IceStorm/weather/Sensor.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) ZeroC, Inc.
+
+#include "../../common/Time.h"
+#include "WeatherStation.h"
+
+#include <Ice/Ice.h>
+#include <IceStorm/IceStorm.h>
+#include <iostream>
+#include <optional>
+#include <random>
+#include <thread>
+
+using namespace ClearSky;
+using namespace std;
+
+int
+main(int argc, char* argv[])
+{
+    // CtrlCHandler must be created before the communicator is created or any other threads are started.
+    Ice::CtrlCHandler ctrlCHandler;
+
+    // Create an Ice communicator. We'll use this communicator to create proxies and manage outgoing connections
+    Ice::CommunicatorPtr communicator = Ice::initialize(argc, argv);
+
+    // Make sure the communicator is destroyed at the end of this scope.
+    Ice::CommunicatorHolder communicatorHolder{communicator};
+
+    // Create a proxy to the IceStorm topic manager.
+    IceStorm::TopicManagerPrx topicManager{communicator, "ClearSky/TopicManager:tcp -p 4061 -h localhost"};
+
+    // Retrieve a proxy to the "weather" topic: we first create a topic with the given name (in case we are the first),
+    // and then retrieve the proxy if the topic was already created.
+    std::optional<IceStorm::TopicPrx> topic;
+    const std::string topicName = "weather";
+    try
+    {
+        topic = topicManager->create(topicName);
+    }
+    catch (const IceStorm::TopicExists&)
+    {
+        topic = topicManager->retrieve(topicName);
+    }
+
+    // The proxy returned by create and retrieve is never null.
+    assert(topic);
+
+    // Create a WeatherStation proxy using the publisher proxy of the topic.
+    std::optional<Ice::ObjectPrx> publisher = topic->getPublisher();
+    // The publisher proxy is never null.
+    assert(publisher);
+    // Create a WeatherStation proxy from the publisher proxy.
+    auto weatherStation = Ice::uncheckedCast<WeatherStationPrx>(*publisher);
+
+    // The proxy returned by IceStorm is a two-way proxy. We can convert it into a oneway proxy if we don't need
+    // acknowledgments from IceStorm.
+    assert(weatherStation->ice_isTwoway());
+
+    // Generate a sensor ID for this weather station.
+    string uuid = Ice::generateUUID();
+    string sensorId = "sensor-" + uuid.substr(uuid.length() - 4);
+
+    cout << sensorId << " reporting. Press Ctrl+C to stop." << endl;
+
+    // Shutdown the sensor on Ctrl-C.
+    std::atomic<bool> running{true};
```

Comment:
I think it would be more appropriate to use a promise and no sleep_for.

Ctrl+C sets the result and clears itself (to avoid double set). The loop waits for the future to become available (for 1 second).

Reply:
That's so much more complicated and unnecessary IMO. There's no treason to use promises for everything.

Reply:
It's not really more complicated and in all other demos the "sleep" gets interrupted by Ctrl+C.

Reply:
Ok, fair enough.